server/handlers/authorization.go

```diff
@@ -18,8 +13,196 @@ import (
 	"golang.org/x/oauth2"
 )
 
-func Authorize(w http.ResponseWriter, r *http.Request) {
-	service := r.Context().Value("Oauthservice").(*svc.Service)
+type AuthorizationHandler interface {
+	Authorize(w http.ResponseWriter, r *http.Request)
+	OAuthToken(w http.ResponseWriter, r *http.Request)
+	Error(w http.ResponseWriter, r *http.Request)
+}
+
+type authorizationHandlerImpl struct {
+	service   svc.ServiceInterface
+	parseHTML func(w http.ResponseWriter, htmlFile string, params map[string]interface{})
+}
+
+func NewAuthorizationHandler(
+	service svc.ServiceInterface,
+	htmlParserFunc func(w http.ResponseWriter, htmlFile string, params map[string]interface{}),
+) AuthorizationHandler {
+	return &authorizationHandlerImpl{
+		service:   service,
+		parseHTML: htmlParserFunc,
+	}
+}
+
+func (a *authorizationHandlerImpl) Authorize(w http.ResponseWriter, r *http.Request) {
+	switch r.Method {
+	case "GET":
+		a.getAuthorizeHandler(w, r)
+	case "POST":
+		a.postAuthorizeHandler(w, r)
+	default:
+		http.Error(w, "Method not allowed", http.StatusMethodNotAllowed)
+	}
+}
+
+func (a *authorizationHandlerImpl) getAuthorizeHandler(w http.ResponseWriter, r *http.Request) {
+	var (
+		clientID          = r.URL.Query().Get("client_id")
+		scopes            = r.URL.Query().Get("scope")
+		redirectURI       = r.URL.Query().Get("redirect_uri")
+		scopeDescriptions = []string{}
+		next              string
+	)
+
+	// use the default scope if none is provided
+	if scopes == "" {
+		scopes = constants.READ_USER_SCOPE
+	}
+
+	// add the scope descriptions
+	for _, scope := range strings.Split(scopes, ",") {
+		scopeDescriptions = append(scopeDescriptions, constants.ScopeDescriptions[scope])
+	}
+
+	// get the client
+	client, err := a.service.GetClient(clientID)
+	if err != nil {
+		log.Println(err)
+		// redirect to the redirect_uri with error
+		http.Redirect(w, r, "/error?opt=invalid_client", http.StatusSeeOther)
+		return
+	}
+
+	// generate the next url
+	uri, err := url.Parse("/authorize")
+	if err != nil {
+		log.Println(err)
```

Comment:
Same here.

Reply:
This log statement is used for debugging purpose, since we dont use any logging library we use println instead but the behavior is same